datalogger/logger.go

```diff
@@ -95,3 +107,122 @@ func (logger *CSVDataLogger[T]) Close() bool {
 	logger.isOpen = false
 	return true
 }
+
+func CreateJSONDataLogger[T any](filename string) (DataLogger[T], error) {
+	fmt.Printf("Creating a JSON data logger: %v!\n", filename)
+	data := make([]T, 0)
+	destination, err := os.Create(filename)
+	if err != nil {
+		return &JSONDataLogger[T]{&sync.Mutex{}, 0, data, true, destination}, err
+	}
+
+	result := JSONDataLogger[T]{&sync.Mutex{}, 0, data, true, destination}
+	return &result, nil
+}
+
+func (logger *JSONDataLogger[T]) LogRecord(record T) {
+	logger.mut.Lock()
+	defer logger.mut.Unlock()
+	logger.recordCount += 1
+	logger.data = append(logger.data, record)
+}
+
+func PrettyPrint(out string, spaces int) string {
+	return fmt.Sprintf("%*s%s\n", spaces, "", out)
+}
+
+// Key Value Pair Print
+func KVPPrint(key string, value any) string {
+	return fmt.Sprintf("\"%s\": \"%v\"", key, value)
+}
+
+func (logger *JSONDataLogger[T]) Export() bool {
+	logger.mut.Lock()
+	defer logger.mut.Unlock()
+	if !logger.isOpen {
+		return false
+	}
+
+	// The rest of the code is written with the assumption that the indexing stays the same
```

Comment:
I think that this is a perfectly valid assumption given the fact that you hold a mutex on the structure while you are printing it, but I just wanted to know exactly what you mean by it!

Reply:
What I meant with this was that I assume the reflect.VisibleFields call would always return fields in the same order, that way when I do
`for ix, v := range visibleFields` on line 194
and use
`data.FieldByIndex(v.index)` on line 195
it always finds the correct data from index. (doesn't have a different type of data in one of the data points)

Reply:
Now I understand! Check out my question below!